04-store-api/starter/controllers/products.js

```diff
@@ -0,0 +1,127 @@
+  // const search = 'ab';
+  // in mongo db docs: $regex Selects documents where values match a specified regular expression.
+  //in this case if above we set const search = 'ab' it will return us all results with letters 'ab' in  name
+  // name: { $regex: search, $options: 'i' },
+  const products = await Product.find({ price: { $gt: 30 } })
+    .sort('price')
+    .select('name price')
+    .limit(10)
+    //skipping 1 item in our response
+    .skip(1);
+  //but if we need to get smth specific: const products = await Product.find({ featured: false }) <--we can add more and combine the coditions: for example: name:'wooden desk'
+
+  //throw new Error('testing async errors package')
+
+  res.status(200).json({ products, nbHits: products.length });
+};
+
+const getAllProducts = async (req, res) => {
+  //here we set up a condition, we will only look for featured caracteristic
+  //we will only show the result if featured (or/and others) is/are passed as a query
+  const { featured, company, name, sort, fields, numericFilters } = req.query;
+  const queryObject = {};
+
+  if (featured) {
+    //here we set up a new property to queryObject called .featured and if the value that user pass for featured in url =true we set it to true and only show results with featured:true, if not - we set it  to 'false' and only show results with festured:false
+    queryObject.featured = featured === 'true' ? true : false;
+  }
+
+  //it will look for the value that user provided in the request, and it will look only amoung previous results (featured:true or featured: false) .
+  //for example, if the query is featured=false and company=ikea there will be 6 results (hits)
+  if (company) {
+    queryObject.company = company;
+  }
+
+  //if the name exists we´ll use regex
+  if (name) {
+    queryObject.name = { $regex: name, $options: 'i' };
+    // or queryObject.name=name means if the name exists i ´ll go with my queryObject, and name property will be = to a name that user provides
+  }
+
+  if (numericFilters) {
+    const operatorMap = {
+      '>': '$gt',
+      '>=': '$gte',
+      '=': '$eq',
+      '<': '$lt',
+      '<=': '$lte',
+    };
+
+    //QUESTION 1 TO MY MENTOR: what does this line 55 means? where did it came from?
+    const regEx = /\b(<|>|>=|=|<|<=)\b/g;
+    //if user passes numericFilters (initially it is a string), then i need to replace it with the value that mongoose understand
+    //so i create a fn filters and pass the regular expression (1st parameter(regEx)) (2nd parameter- callback function)
+    
+    // in operatorMap[match] we get access to the value that user provides(match), it could be > , =, etc. and then callback translate in to the mongoose language($...)
+    //so now we can get the result
+    //QUESTION 2 TO MY MENTOR:`-${operatorMap[match]}-` what does this symbols (-)do in this case ?
+    let filters = numericFilters.replace(
+      regEx,
+      (match) => `-${operatorMap[match]}-`
+    );
```

Comment:
Here we're calling the [String.prototype.replace](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace) method, passing in the regex as the first argument and a function as the second argument. The function will be called every time the regex finds a match in the string `numericFilters`.

In this case, the function we're using says: "when a match is found, lookup the match in the `operatorMap` object, and replace it in a string with the format `-<result>-`.

So if I had the following as my numericFilters query parameter, `price>30,rating=4.5`, then the regex would find two matches:

![](https://user-images.githubusercontent.com/9094098/230906944-c7ee910b-92f8-4e0d-8612-38a2240d77bc.png)

And then it would replace each of those matches (so the `>` and the `=`), with the corresponding string from `operatorMap` which is the Mongo operator that's the equivalent of the human-friendly character that was passed in.

Recording of trying those steps in node: https://asciinema.org/a/votZugxuNEMhQh1CprjZAM8ve

Reply:
The instructor arbitrarily picked `-` but it could have been any character. We just want to pick something to use as a separator so that we can easily split up this text later on line 74. And we want to pick a character that's not used elsewhere in the `numericFilters` string already so that we know it only occurs between the name of the field we're filtering on (eg: `price`) and the numeric value we're filtering by (eg: `30`).

Reply:
@akosasante thank you! I´ve never got such a great feedback on my work and questions!

Reply:
Happy to help! :smile: